### ISSUE TITLE
No Speedup with OpenMP

### ISSUE THREAD

**Issue by phildue** [OctoMap/octomap#243]
Hi,

I want to speedup octomap using OpenMP. I created a small test application where I simply insert a point cloud using insertPointCloud(). 

However, even for 1000 000 points I can not really measure an improve in performance when using OpenMP.

I verified that compiling and linking worked correctly and I am sure that the OpenMP code is running properly. Do you have an explanation why I can't measure any speedup? Or is there somewhere a small example that demonstrates the parallelization?

Thanks already!

**Comment by ahornung** [comment]
The speedup depends stongly on the number of threads (set with environment variable OMP_NUM_THREADS), in my experience it is greatest between 2-4 threads.

Parallelization is used at the following lines for the ray computation. The actual node update is protected as "critical", as the actual octree data structure it not multi-threading save. 
https://github.com/OctoMap/octomap/blob/bf3b3949a99c13ef34d11dc8bc23dcd6680a9f3c/octomap/include/octomap/OccupancyOcTreeBase.hxx#L120

**Comment by peci1** [comment]
Well, because of the critical sections in the code, the multithreading doesn't actually help a lot. The performance boost given by multiple threads is eaten by the cost of managing multiple threads and locking the critical sections.

What I tried is dividing the `free_cells` and `occupied_cells` to several pieces, each for one thread (as is done with `keyrays`), which allowed to remove the critical sections and even a profiler showed much better utilization of the threads. However, after the parallelized for loop, a merge of these sets is needed, and that showed to be a bottleneck (over 2 seconds for 15 million points). If somebody figured out a more efficient way to merge the sets, then this could be a real speedup.

One possible way for improvement would be not iteraring the points one-by-one, but in chunks (e.g. 10 000 points a time). That would allow threads to spend some time computing the raytracing, while other threads could be extending the sets inside critical sections at the same time. But an optimal chunk size would probably be use-case-dependent.

**Comment by peci1** [comment]
Anyways, there's another bottleneck, when `updateNode()` calls are made, and that one takes approximately the same amount of time as the raytracing itself. So even if you managed to speed up the raytracing using OpenMP, you would cut at best a half of the run time.

Is there hope the `updateNode` calls could be parallelized? I know about the `lazy_eval` parameter, which helps a bit, but it is still too slow.

**Comment by ahornung** [comment]
Yes, it's a hard problem to parallelize in the end because all information needs to end up in one consistent tree structure. It could be parallelized by finding "split paths" into two different tree branches as early as possible and delaying a consistent update (similar to lazy_eval?) as late as possible.

All in all the gain could be highly dependent on the used data (point distribution), as with the ray computation before.
